### DISCUSSION TITLE
Horizontal Scrolling ItemsControl  (with mouse)

### DISCUSSION THREAD

**Discussion by DominicMaas** [AvaloniaUI/Avalonia#9011] [Q&A] [answered]
I cannot be the only one who has wanted to do this using Avalonia, I just cannot seem to understand how to make a horizontal scrolling items control.

```xml
<ScrollViewer HorizontalScrollBarVisibility="Visible" VerticalScrollBarVisibility="Disabled">
    <ItemsControl>
        <ItemsControl.ItemsPanel>
            <ItemsPanelTemplate>
                <StackPanel Orientation="Horizontal" />
            </ItemsPanelTemplate>
        </ItemsControl.ItemsPanel>
    </ItemsControl>
</ScrollViewer>
```

This displays as expected, and I can use the mouse pointer to move side to side, but the mouse wheel does not work. If I hold done shift it works. 

Is this supported? Otherwise, what is a good path to implement this? I looked at an attached property (like WPF), but not picking up the `PointerWheelChangedEvent` on the scrollviewer.

WinUI supports this quite nicely with just the following:

```xml
<Setter Property="HorizontalScrollBarVisibility" Value="Auto" />
<Setter Property="HorizontalScrollMode" Value="Enabled" />
<Setter Property="VerticalScrollMode" Value="Disabled" />
<Setter Property="IsHorizontalRailEnabled" Value="True" />
<Setter Property="IsVerticalRailEnabled" Value="False" />
```

**Avalonia**: 0.10.18
**OS**: Ubuntu 22.04.1 LTS

Thanks

**Comment by maxkatz6** [comment, marked as answer]
No. Currently it's not supported as you can find here in the code https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs#L438

> but not picking up the PointerWheelChangedEvent on the scrollviewer

It's handled on the ScrollContentPresenter and you would need to register event handled with Tunnel strategy. Something like `presenter.AddHandler(Event, Handler, Tunnel);`. "Tunnel" events work similarly how "Preview" events work in UWP.

> WinUI supports this quite nicely with just the following:

I don't think railing is related here.
And I suppose UWP automatically flips wheel direction, when there is no vertical scroll bar enabled.

**Reply by maxkatz6** [reply]
We can have this feature implemented, if somebody wants to help.
But I am not sure if implicit behavior would be the best here. Additional property to disable it maybe? Or are developers used to this automatic wheel flipping?

**Reply by DominicMaas** [reply]
Thanks for the hint with the "Tunnel" parameter! Attached that event to the scrollviewer, am able to simply override the scroll behaviour.

**Comment by kyuranger** [comment]
I found a stupid way to achieve this based on the answer in [How to enable horizontal scrolling with mouse?](https://stackoverflow.com/a/53228194/12966640):
```
TestScrollViewer.PointerWheelChanged += (sender, e) => {
    if (e.Delta.Y > 0)
    {
        TestScrollViewer.PageLeft();
    }
    else
    {
        TestScrollViewer.PageRight();
    }
};
```

It works now.
